### PR TITLE
add bismark/methylation_extractor

### PR DESCRIPTION
**Pull request by phue**
## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `<SOFTWARE>.version.txt` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

**Comment by phue** [comment]
Today I learned that gzip does not produce deterministic results because it stores timestamps in the header. That's why the md5sums will keep changing.

Even if we omit the  `--gzip` flag from the process, the `bedGraph` and `coverage` outputs will be compressed by default, so that is not really an option.
Maybe there is a way to fake a timestamp in Github actions?
Otherwise I would just remove the md5sum checks for the compressed files, unless someone has an idea?

File: software/bismark/methylation_extractor/main.nf [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    } else {
+        container "quay.io/biocontainers/bismark:0.23.0--0"
+    }
+
+    input:
+    tuple val(meta), path(bam)
+    path index
+
+    output:
+    tuple val(meta), path("*.bedGraph.gz")         , emit: bedGraph
```
Comment (drpatelh) [inline software/bismark/methylation_extractor/main.nf:26]:
```suggestion
    tuple val(meta), path("*.bedGraph.gz")         , emit: bedgraph
```

File: software/bismark/methylation_extractor/main.nf [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    script:
+    def seqtype  = meta.single_end ? '-s' : '-p'
+    def software = getSoftwareName(task.process)
+    """
+    bismark_methylation_extractor \\
+        --bedGraph \\
+        --counts \\
+        --gzip \\
+        --report \\
```
Comment (drpatelh) [inline software/bismark/methylation_extractor/main.nf:41]: Should these be coming in via `options.args` instead?

File: software/bismark/methylation_extractor/main.nf [hunk 3]
```diff
@@ -0,0 +1,48 @@
+// Import generic module functions
+include { initOptions; saveFiles; getSoftwareName } from './functions'
+
+params.options = [:]
+def options    = initOptions(params.options)
```
Comment (drpatelh) [inline software/bismark/methylation_extractor/main.nf:5]:
```suggestion
options        = initOptions(params.options)
```

File: software/bismark/methylation_extractor/meta.yml [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        type: dir
+        description: Bismark genome index directory
+        pattern: "BismarkIndex"
+output:
+    - meta:
+        type: map
+        description: |
+            Groovy Map containing sample information
+            e.g. [ id:'test', single_end:false ]
+    - bedGraph:
```
Comment (drpatelh) [inline software/bismark/methylation_extractor/meta.yml:62]:
```suggestion
    - bedgraph:
```

**Review by drpatelh — CHANGES_REQUESTED** [review]
Looks great @phue ! Couple of changes.

File: software/bismark/methylation_extractor/main.nf [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+process BISMARK_METHYLATION_EXTRACTOR {
+    tag "$meta.id"
+    label 'process_high'
+    publishDir "${params.outdir}",
+        mode: params.publish_dir_mode,
+        saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:getSoftwareName(task.process), publish_id:meta.id) }
+
+    conda (params.enable_conda ? "bioconda::bismark==0.23.0" : null)
```
Comment (drpatelh) [inline software/bismark/methylation_extractor/main.nf:14]:
```suggestion
    conda (params.enable_conda ? "bioconda::bismark=0.23.0" : null)
```

**Comment by drpatelh** [comment]
> Today I learned that gzip does not produce deterministic results because it stores timestamps in the header. That's why the md5sums will keep changing.

No way?!  Does that go for any zipped file? Assuming you have ruled out this isn't down to the core file changing?

**Comment by phue** [comment]
> > Today I learned that gzip does not produce deterministic results because it stores timestamps in the header. That's why the md5sums will keep changing.
> 
> No way?! Does that go for any zipped file? Assuming you have ruled out this isn't down to the core file changing?

When I extract the files, checksums match, so gzip is the culprit here

File: software/bismark/methylation_extractor/main.nf (hunk 2 above)
Comment (phue) [reply software/bismark/methylation_extractor/main.nf:41]: I *think* these are sensible defaults. By default, the tool produces output per strand so the `--bedGraph` is useful to consolidate this into one file. Also the output can get really massive so it is sensible to put `--gzip`
And `--report` is something that `bismark_summary` and ultimately `MultiQC` needs

Comment (FelixKrueger) [reply software/bismark/methylation_extractor/main.nf:41]: Just to clarify, `--report` is on by default, but doesn't hurt to specify explicitly. 

The same goes for `--counts`: the coverage files does have the counts, the `bedGraph` file doesn't, and both files are produced for each input if `--bedGraph` is given.

Comment (drpatelh) [reply software/bismark/methylation_extractor/main.nf:41]: Cool beans!

**Review by drpatelh — APPROVED** [review]
Gracias!

**Pull request merged**